src/nanoarrow/nanoarrow_device_cuda.c

```diff
@@ -392,6 +417,251 @@ static ArrowErrorCode ArrowDeviceCudaArrayMove(struct ArrowDevice* device_src,
+  // Currently no attempt to minimize the amount of memory copied (i.e.,
+  // by applying offset + length and copying potentially fewer bytes)
+  dst->length = src->length;
+  dst->offset = src->offset;
+  dst->null_count = src->null_count;
+
+  for (int i = 0; i < NANOARROW_MAX_FIXED_BUFFERS; i++) {
+    if (src->layout.buffer_type[i] == NANOARROW_BUFFER_TYPE_NONE) {
+      break;
+    }
+
+    struct ArrowBufferView src_view = src->buffer_views[i];
+    int64_t buffer_size = src_view.size_bytes;
+
+    // No need to initialize a buffer of zero bytes
+    if (buffer_size == 0) {
+      continue;
+    }
+
+    // buffer_size will be -1 if it is unknown (i.e., string/binary
+    // data buffer when copying from CUDA -> CPU)
+    if (buffer_size < 0) {
+      (*unknown_buffer_size_count)++;
+      continue;
+    }
+
+    struct ArrowBuffer* buffer_dst = ArrowArrayBuffer(dst, i);
+
+    // Don't initialize a buffer if we have already started the copy.
+    // This check works because the buffer size is set synchronously
+    // when the copy operation is launched.
+    if (buffer_dst->size_bytes > 0) {
+      continue;
+    }
+
+    // Otherwise, initialize the buffer
+    NANOARROW_RETURN_NOT_OK(ArrowDeviceCudaBufferInitAsync(
+        device_src, src_view, device_dst, buffer_dst, hstream));
+  }
+
+  // Recurse for children
+  for (int64_t i = 0; i < src->n_children; i++) {
+    NANOARROW_RETURN_NOT_OK(ArrowDeviceCudaArrayViewCopyBuffers(
+        device_src, src->children[i], device_dst, dst->children[i], hstream,
+        unknown_buffer_size_count));
+  }
+
+  // ...and for dictionary
+  if (src->dictionary != NULL) {
+    NANOARROW_RETURN_NOT_OK(ArrowDeviceCudaArrayViewCopyBuffers(
+        device_src, src->dictionary, device_dst, dst->dictionary, hstream,
+        unknown_buffer_size_count));
+  }
+
+  return NANOARROW_OK;
+}
+
+static ArrowErrorCode ArrowDeviceCudaCreateStream(struct ArrowDevice* src,
+                                                  struct ArrowDevice* dst, CUstream* out,
+                                                  struct ArrowError* error) {
+  struct ArrowDeviceCudaPrivate* private_data;
+
+  if (dst->device_type == ARROW_DEVICE_CUDA) {
+    private_data = (struct ArrowDeviceCudaPrivate*)dst->private_data;
+  } else if (src->device_type == ARROW_DEVICE_CUDA) {
+    private_data = (struct ArrowDeviceCudaPrivate*)src->private_data;
+  } else {
+    return NANOARROW_OK;
+  }
+
+  NANOARROW_CUDA_RETURN_NOT_OK(cuCtxPushCurrent(private_data->cu_context),
+                               "cuCtxPushCurrent", NULL);
+  CUresult err = cuStreamCreate(out, CU_STREAM_NON_BLOCKING);
+  CUcontext unused;
+  NANOARROW_CUDA_ASSERT_OK(cuCtxPopCurrent(&unused));
+  NANOARROW_CUDA_RETURN_NOT_OK(err, "cuStreamCreate", error);
+
+  return NANOARROW_OK;
+}
+
+static void ArrowDeviceCudaDestroyStream(CUstream hstream) {
+  if (hstream != NANOARROW_CUDA_DEFAULT_STREAM) {
+    NANOARROW_CUDA_ASSERT_OK(cuStreamDestroy(hstream));
+  }
```

Comment:
In addition to the default stream there's also the legacy default stream and per thread default stream (see `CU_STREAM_LEGACY` and `CU_STREAM_PER_THREAD` [here](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES))